### PR TITLE
Add a module to gray-code and synchronize multibit signals across a c…

### PR DESCRIPTION
**Pull request by ben-k**
…lock interface

### CODE REVIEW

**Comment by colinschmidt** [comment]
I ended up just making the changes myself.

File: src/main/scala/Util.scala [hunk 1]
```diff
@@ -78,3 +78,33 @@ class PutSeqDriver(val s: Seq[Tuple2[BigInt,Int]])(implicit p: Parameters) exten
...
+class WordSync[T <: Data](gen: T, lat: Int = 2) extends Module {
+  val size = gen.cloneType.fromBits(UInt(0)).asUInt().getWidth
+  val io = new Bundle {
+    val in = gen.cloneType.flip
+    val out = gen.cloneType
+    val tx_clock = Clock(INPUT)
+  }
+  val bin2gray = Module(new BinToGray(gen,io.tx_clock))
+  val out_gray = ShiftRegister(bin2gray.io.gray, lat)
+  io.out := gen.cloneType.fromBits((0 until size).map{ out_gray.asUInt >> UInt(_) }.reduceLeft(_^_))
```
Comment (ben-k) [inline src/main/scala/Util.scala]: The unneeded XORs will just get optimized away by the synthesis tool, no?

I'm not averse to bumping Chisel, though.

File: src/main/scala/Util.scala [hunk 2]
```diff
@@ -78,3 +78,33 @@ class PutSeqDriver(val s: Seq[Tuple2[BigInt,Int]])(implicit p: Parameters) exten
 
 }
 
+// Use gray coding to safely synchronize a word across a clock crossing.
+// This should be placed in the receiver's clock domain.
+class WordSync[T <: Data](gen: T, lat: Int = 2) extends Module {
+  val size = gen.cloneType.fromBits(UInt(0)).asUInt().getWidth
+  val io = new Bundle {
+    val in = gen.cloneType.flip
```
Comment (ben-k) [inline src/main/scala/Util.scala]: I have trouble keeping track of what chisel3 newfangled-ness is actually implemented (and available in our version)...

Comment (colinschmidt) [inline src/main/scala/Util.scala]: Yeah I got the right version of chisel just by merge from the hwacha branch of rocket-chip (but not before)

**Comment by ben-k** [comment]
Is there a reason we're not merging this into master?

**Comment by colinschmidt** [comment]
I'm happy to merge in into master if this is the only branch in use. @zhemao Does craft use this?

**Comment by zhemao** [comment]
Sorry, I wasn't sure what we were doing with this. I'm pretty sure Craft is using this. So is fpga-zynq. Are these changes compatible with the chisel versions being used in those repos?

**Comment by colinschmidt** [comment]
I honestly have no idea what version of chisel other people are using. This works as of https://github.com/ucb-bar/chisel3/commit/2a4d4f0ee9fc924ee9ff8a12c7a04a4c8e9a106b

**Comment by zhemao** [comment]
I'm going to bump chisel3 and firrtl in the project-template-compat branch of rocket-chip. That will allow fpga-zynq and project-template to use this change.

**Pull request merged**